content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
+            
+            /* forward return to users */
+            for (auto& U : I.uses()) {
+                User* user = U.getUser();
+                user->setOperand(U.getOperandNo(), ret);
+            }
+            
+            /* mark instruction for deletion */
+            workList.push(&I);
+            I.dropAllReferences();
+        }
+        
+        /* delete replaced instructions */
+        while (!workList.empty()) {
+            workList.front()->eraseFromParent();
+            workList.pop();
+        }
+    }
+```
+Key limitations to note, runtime library has to be co-designed with the function declarations 
+baked into the llvm pass and of course, the instruction selection algorithm can get very complex
+depending on how interesting the task is.
+
+I wish to direct your attention also to the granularity of the replacements, the runtime functions
+are essentially overloading the llvm instructions here. This granularity is not particulary useful for
+loosely-coupled accelerators, as communication between the accelerator and the CPU is at the highest.
+This is reflected in the performance section.
+
+## Replacing functions
+In contrast to the instruction level, functions are a substantially coarser-grained abstraction 
+for an accelerator to replace. Even a loosely-coupled accelerator can tradeoff the communication latencies
+by efficient computation. 
+
+```c
+    /* create the function call from runtime library */
+    LLVMContext& Ctx = F.getContext();
+    FunctionCallee rtFunc;
+    rtFunc = F.getParent()->getOrInsertFunction(
+        "func_replace", 
+        Type::getVoidTy(Ctx),
+        Type::getDoublePtrTy(Ctx), Type::getDoublePtrTy(Ctx), Type::getDoublePtrTy(Ctx)
+    );
+    // function call name and argument types have to be known
+
+    /* Add runtime function selection here */
+    
+    /* End of selection algorithm */
+        
+    queue <BasicBlock*> workList; // collects replaced function within iterator
+    /* mark basic blocks for deletion */
+    for (auto& B : F) {
+        for (auto& I : B) {
+            Instruction* ins = &I;
+            ins->dropAllReferences();
+        }
+        workList.push(&B);
+    }
+    
+    /* delete basic blocks in the function being replaced */
+    while (!workList.empty()) {
+        workList.front()->eraseFromParent();
+        workList.pop();
+    }
+    
+    /* collect function arguments
+    int i = 0;
+    Value* args[3];
+    for(auto arg = F.arg_begin(); arg != F.arg_end(); ++arg) {
+            args[i] = arg;
+            i++;
+    }
+    // requires knowlegde about arguments
+
+    /* create new basic block */ 
+    BasicBlock *pb = BasicBlock::Create(Ctx, "simple", &F);
+    
+    /* create the runtime function call */
+    IRBuilder<> builder(pb);
+    builder.CreateCall(rtFunc, args);
+    
+    /* create a return instruction from the return value */
+    Instruction *retInst = ReturnInst::Create(Ctx, pb);
+```
```

Comment:
Add an explanation here to say what the code does. Looks like it empties out a function body and replaces with a call? Showing pseudocode before and after the transformation would help.